src/ucs/datastruct/bitmap.h

```diff
@@ -0,0 +1,170 @@
+ * of 64-bit unsigned long integers.
+ * 
+ * @param _length Bits number in the bitmap
+ */
+#define _UCS_BITMAP_DECLARE_TYPE(_length) \
+    typedef struct { \
+        uint64_t bits[_UCS_BITMAP_WORD_INDEX(_length)]; \
+    } ucs_bitmap_t(_length);
+
+
+/**
+ * Expands to bitmap type definition
+ *
+ * @param _length Bits number (as passed to UCS_BITMAP_DECLARE_TYPE)
+ *
+ * Example:
+ *
+ * @code{.c}
+ * ucs_bitmap_t(64) my_bitmap;
+ * @endcode
+ */
+#define ucs_bitmap_t(_length) \
+    _UCS_BITMAP_IDENTIFIER(_length, _t)
+
+
+#define _UCS_BITMAP_IDENTIFIER(_name, _suffix) \
+    UCS_PP_TOKENPASTE(ucs_bitmap_, UCS_PP_TOKENPASTE(_name, _suffix))
+
+
+_UCS_BITMAP_DECLARE_TYPE(64)
+_UCS_BITMAP_DECLARE_TYPE(128)
+_UCS_BITMAP_DECLARE_TYPE(256)
+
+
+/**
+ * Get bit value
+ * 
+ * @param _bitmap Bitmap to get from
+ * @param _index Bit index
+ * 
+ * @return Bit value (0 or 1)
+*/
+#define UCS_BITMAP_GET(_bitmap, _index) \
+    (!!((_bitmap)->bits[_UCS_BITMAP_WORD_INDEX(_index)] & \
+    UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index))))
+
+
+/**
+ * Set a bit
+ * 
+ * @param _bitmap Bitmap to set in
+ * @param _index Bit index
+ */
+ #define UCS_BITMAP_SET(_bitmap, _index) \
+    ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] |= \
+    (UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index)));
+
+
+/**
+ * Unset a bit
+ * 
+ * @param _bitmap Bitmap to unset in
+ * @param _index Bit index
+ */
+ #define UCS_BITMAP_UNSET(_bitmap, _index) \
+     ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] &= \
+     ~((UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index))));
+
+
+/**
+ * Clear a bitmap by setting all bits to zero
+ * 
+ * @param _bitmap Bitmap to clear
+ * 
+ * @return Value of the bit (0 or 1)
+ */
+#define UCS_BITMAP_CLEAR(_bitmap) \
+    (memset((_bitmap)->bits, 0, sizeof((_bitmap)->bits)))
+
+
+/**
+ * Iterate over all set bits of a given bitmap
+ * 
+ * @param _bitmap Bitmap to iterate on
+ * @param _index Bit index (global offset - relative to the whole bitmap)
+ * @param _word Index of the 64-bit word in which the bit is set
+ * @param _index_in_word Index of the bit in its 64-bit word
+ */
+#define UCS_BITMAP_FOR_EACH_BIT(_bitmap, _index, _word, _index_in_word) \
+    for (_word = 0, _index = 0, _index_in_word = 0; \
+         _word < ucs_static_array_size((_bitmap)->bits); _index_in_word = 0, _word++) \
+        ucs_for_each_bit (_index_in_word, (_bitmap)->bits[_word]) \
+            if ((bool)(_index = _word * 64 + _index_in_word) || true)
+
+
+/**
+ * Perform bitwise NOT of a bitmap
+ * 
+ * @param _bitmap Bitmap to negate
+ */
+#define UCS_BITMAP_NOT(_bitmap) \
+        for (int _word = 0; _word < ucs_static_array_size((_bitmap)->bits); \
+             _word++) \
+            (_bitmap)->bits[_word] = ~(_bitmap)->bits[_word];
+
+
+/**
+ * Perform bitwise AND of 2 bitmaps, storing the result in the first one
+ * 
+ * @param _bitmap1 First bitmap
+ * @param _bitmap1 Second bitmap
+ */
+#define UCS_BITMAP_AND(_bitmap1, _bitmap2) \
```

Comment:
maybe add generic binary operator macro which receives operator (and/or/xor) and 2 bitmaps?